### ISSUE TITLE
Negative-Jacobian tests in TensorMechanics

### ISSUE THREAD

**Issue by WilkAndy** [idaholab/moose#9740]
### Description of the enhancement or error report

Add two tests to TensorMechanics:

 - a test that inverts an element but because it use_displaced_mesh=false (and small strain) it won't produce an error

 - an identical test that triggers an error because it use_displaced_mesh=true

### Rationale for the enhancement or information for reproducing the error

To prevent regression back to a version of moose that produces these super annoying errors in the non-displaced mesh case

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

None.

Tag @bwspenc

**Comment by permcody** [comment]
It would be even better if this wasn't a mechanics test but that's where I've failed. You should be able turn on displaced mesh without mechanics and displace the mesh by the value given in any old variable (aux or nonlinear). This test really should be in the framework, but I'll settle for anything at this point.

**Comment by dschwen** [comment]
```
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 10
  displacements = disp_x
[]

[AuxVariables]
  [./disp_x]
    [./InitialCondition]
      type = FunctionIC
      function = -x
  [../]
[]

[Variables]
  [./u]
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
    use_displaced_mesh = true
  [../]
[]

[Executioner]
  type = Steady
[]
```
You're welcome!

**Comment by bwspenc** [comment]
Thanks, I see that one does indeed fail. I also made that into a 2D problem using the Steady Executioner as you did, and it also fails, which is good.  However, I made it into a transient problem that ramps up the displacement, and it happily runs even when the element is inverted. This problem should fail:

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  displacements = 'disp_x disp_y'
[]

[AuxVariables]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
[]

[AuxKernels]
  [./disp_x]
    variable = disp_x
    type = FunctionAux
    function = '-x*t'
  [../]
[]

[Variables]
  [./u]
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
    use_displaced_mesh = true
  [../]
[]

[Executioner]
  type = Transient

  dt = 0.1
  end_time = 1.2
[]

[Outputs]
  exodus = true
[]
```

**Comment by permcody** [comment]
So do these problems fail now, or just on an older MOOSE? Andy seems to think his negative Jacobian error is fixed. I suppose it's possible that his error is fixed but if you intentionally invert elements you still get the error.

**Comment by dschwen** [comment]
Mine fails now

**Comment by bwspenc** [comment]
My test does not fail either with or without the `use_displaced_mesh` option. I think it should fail when `use_displaced_mesh=true`, but probably not if it is false.

**Comment by WilkAndy** [comment]
OK, this is interesting.  I cannot get @bwspenc 's test to fail if ```use_displaced_mesh=false```.  That is the behaviour i'd like to see in my production simulations, by the way.

However, with ```use_displaced_mesh=true``` , @bwspenc 's test fails if dt=0.2 or dt=1/2^n (eg 0.125, 0.25 or 0.5), but does not fail if dt=0.1, dt=0.3, dt=0.4.

Over to you guys for a while....

**Comment by permcody** [comment]
Yeah - welcome to the pain train.

**Comment by WilkAndy** [comment]
Has anyone any ideas why this failure/non-failure is happening with the negative Jacobian?  I might have some time to look at this today, but if you could me some leads you might save me a lot of time!

**Comment by WilkAndy** [comment]
Oh, I've been continuously working on this for the past 50 days and i've finally figured it out.  The problem with @dschwen and @bwspenc 's files above is that they are 1D.  The only way libmesh gives a "negative Jacobian" error in 1D is when the Jacobian is exactly zero, which always occurs in @dschwen 's case, and only occurs in @bwspenc 's case when dt is chosen in the special ways as i note above.  I'll change the above examples to 3D and submit a PR.   (Just to save you, reader, a heart attack: i was joking about the 50 days!)

**Comment by bwspenc** [comment]
@WilkAndy My case was actually 2D. I'm glad to see you got it to fail in that 3D case, which is very similar to these, but I'm still a little puzzled...

**Comment by bwspenc** [comment]
I'm reopening this issue until we having a working 2D variant of this test.

**Comment by WilkAndy** [comment]
Yep, this should be simple to investigate.  The Jacobian code is in ```libmesh/src/fe/fe_map.C```.  It's the weekend here now, but i should get to this next week.
